app/models/reviewable.rb

```diff
@@ -356,6 +356,10 @@ def perform(performed_by, action_id, args = nil)
 
     validate_action!(guardian, action_id, perform_method, args)
 
+    # Bundle needs to be determined before the action is performed, as bundles
+    # are dynamic based on the state of the reviewable.
+    action_bundle_id = get_action_bundle_id(guardian, action_id, args)
+
     result = nil
     update_count = false
     Reviewable.transaction do
```

Comment:
I'm not wild about how big this transaction is now, but given how each step depends on the next, I don't think it's possible to simplify any further without changing existing behaviour.